### PR TITLE
Feature/forall exists

### PR DESCRIPTION
**Pull request by vpipkt**
Naming of functions TBD. Considerations:

 * forall and exists are pretty common key words in scala. For scala API users if we use those terms we want the mental model of the function to match to what's happening
 * `all` is a keyword in Python
 * in numpy there are methods on arrays: `a.all()` and `a.any()` that do similar to the intent here, but with less explicit handling of nodata
 * Sometimes we prefix a column by `tile_` if the name would otherwise be pretty common (tile_sum, tile_max). Maybe this should be `tile_for_all` or `cells_for_all`?
 * Another thought: `is_no_data_tile` does something like for_all proposed here, testing the cells are all nodata. If we also had a `undefined` function; `is_no_data_tile(t)` would be equivalent to `for_all(undefined(t))`

https://s22s.myjetbrains.com/youtrack/issue/RF-18

Issue about defined / undefined : https://s22s.myjetbrains.com/youtrack/issue/AEP-328

### CODE REVIEW

**Comment by vpipkt** [comment]
Stacks on top of https://github.com/s22s/rasterframes/pull/38

File: core/src/main/scala/astraea/spark/rasterframes/RasterFunctions.scala [hunk 1]
```diff
@@ -160,6 +160,15 @@ trait RasterFunctions {
   def is_no_data_tile(tile: Column): TypedColumn[Any, Boolean] =
     IsNoDataTile(tile)
 
+  /** Returns true if any cells in the tile are true (non-zero and not nodata). */
+  def exists(tile: Column): TypedColumn[Any, Boolean] = Exists(tile)
+
+  /** Returns true if any cells in the tile are true (non-zero and not nodata). */
+  def any(tile: Column): TypedColumn[Any, Boolean] = Exists(tile)
+
```
Comment (metasim) [inline core/src/main/scala/astraea/spark/rasterframes/RasterFunctions.scala]: Why the aliasing? I feel we should just have one function name for the one behavior.

File: core/src/main/scala/astraea/spark/rasterframes/expressions/package.scala [hunk 2]
```diff
@@ -93,6 +93,10 @@ package object expressions {
     registry.registerExpression[DataCells]("rf_data_cells")
     registry.registerExpression[NoDataCells]("rf_no_data_cells")
     registry.registerExpression[IsNoDataTile]("rf_is_no_data_tile")
+    registry.registerExpression[Exists]("rf_exists")
+    registry.registerExpression[Exists]("rf_any")
```
Comment (metasim) [inline core/src/main/scala/astraea/spark/rasterframes/expressions/package.scala]: Ditto on aliasing.

File: core/src/main/scala/astraea/spark/rasterframes/expressions/package.scala [hunk 3]
```diff
@@ -93,6 +93,10 @@ package object expressions {
...
+    registry.registerExpression[ForAll]("rf_for_all")
+    registry.registerExpression[ForAll]("rf_all")
```
Comment (metasim) [inline core/src/main/scala/astraea/spark/rasterframes/expressions/package.scala]: Ditto. Also, we're introducing a name (`all`) that doesn't exist in the JVM library.

File: core/src/main/scala/astraea/spark/rasterframes/expressions/tilestats/ForAll.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    var (c, r) = (0, 0)
+    while (r < tile.rows) {
+      while(c < tile.cols) {
+        if(tile.cellType.isFloatingPoint) { if(! doubleValueIs(tile.getDouble(c, r))) return false }
+        else { if(! intValueIs(tile.get(c, r))) return false }
+        c += 1
+      }
+      c = 0; r += 1
+    }
```
Comment (metasim) [inline core/src/main/scala/astraea/spark/rasterframes/expressions/tilestats/ForAll.scala]: Feels like there's some code duplication going on; in two areas, one subtle, the other more obvious. 

The subtle one is that there should be a single place in the system where we define what is interpreted as `true` and `false` in the context of a cell (just like GeoTrellis does with `isNoData`. I'd suggest putting functions in the `rasterframes` package object. Maybe something like one of these:

* `cellAsBoolean(v: {Int|Double})`
* `interpretedAsBoolean(v: {Int|Double})`
* `isTrueish(v: {Int|Double})` (or `isTruthy` ;-) )
* `isCellTrue(v: {Int|Double})`
* `isDataTrue(v: {Int|Double})`


Remember that in Scala you have function overloading, so you don't need to prefix the function with `int` or `double`.

The other duplication area is that `op` implementations seem very similar in structure, but I admit that the short circuit evaluation you're doing may make it hard to refactor.

File: core/src/main/scala/astraea/spark/rasterframes/expressions/tilestats/ForAll.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  def apply(tile: Column): TypedColumn[Any, Boolean] = new Column(ForAll(tile.expr)).as[Boolean]
+
+  def op(tile: Tile): Boolean = {
+
+    def doubleValueIs(d: Double): Boolean =  isData(d) & d != 0.0
+    def intValueIs(i: Int): Boolean =  isData(i) & i != 0
+
+    var (c, r) = (0, 0)
+    while (r < tile.rows) {
+      while(c < tile.cols) {
```
Comment (metasim) [inline core/src/main/scala/astraea/spark/rasterframes/expressions/tilestats/ForAll.scala]: Use the `cfor` construct as is customary in GeoTrellis and elsewhere in RF.

File: pyrasterframes/python/tests/PyRasterFramesTests.py [hunk 6]
```diff
@@ -216,6 +216,23 @@ def test_resample(self):
 
         self.assertTrue(result == 1)  # short hand for all values are true
 
+    def test_exists_for_all(self):
```
Comment (metasim) [inline pyrasterframes/python/tests/PyRasterFramesTests.py:226]: Python tests 🙇

File: docs/src/main/tut/reference.md [hunk 7]
```diff
@@ -277,7 +277,7 @@ _Python_:
 
 _SQL_: `rf_is_no_data_tile`
  
-Returns true if `tile` contains only nodata. By definition returns false if cell type does not support nodata.
+Returns true if `tile` contains only nodata. By definition returns false if cell type does not support nodata. To count nodata cells or data cells, see @ref:[`no_data_cells`](reference.md#no-data-cells), @ref:[`data_cells`](reference.md#data-cells), @ref:[`agg_no_data_cells`](reference.md#agg-no-data-cells), @ref:[`agg_data_cells`](reference.md#agg-data-cells), @ref:[`agg_local_no_data_cells`](reference.md#agg-local-no-data-cells), and @ref:[`agg_local_data_cells`](reference.md#agg-local-data-cells).
```
Comment (metasim) [inline docs/src/main/tut/reference.md]: Keep a lookout for other instances where the distinction between `is_no_data_tile` can be explained (i.e. zero is treated specially in these boolean evaluation cases). I can see this being quite confusing.

**Review by metasim — CHANGES_REQUESTED** [review]
Generally looks very good, as always. Would like to get this "just right" so we never have to address it again, so I think it's worth addressing the issues I've enumerated, and for us to get the naming on the mark. Some off-the-cuff ideas (none are that great):

## Instead of `exists` or `any`

* `has_true_cell`
* `has_true_data`
* `has_true`

## Instead of `all` or `for_all`

* `all_cells_true`
* `all_values_true`
* `all_true`

However, if we consider renaming all columnar functions to have the `rf_` prefix, maybe this issue goes away?

**Comment by metasim** [comment]
Rebase on https://github.com/s22s/rasterframes/pull/39 and lets see how the naming works out. I suspect the issues go away, and `rf_any` and `rf_all` will seem natural.

**Comment by vpipkt** [comment]
I decided it was too much hassle to rebase... But I also don't want to loose the comments here. Should i force push?

File: core/src/main/scala/astraea/spark/rasterframes/expressions/tilestats/ForAll.scala (hunk 4 above)
Comment (vpipkt) [reply core/src/main/scala/astraea/spark/rasterframes/expressions/tilestats/ForAll.scala]: I have no real vision on how to reduce apparent duplication of the short circuited loops here. But point very much taken on the other.

File: docs/src/main/tut/reference.md (hunk 7 above)
Comment (vpipkt) [reply docs/src/main/tut/reference.md]: I added some hopefully clarifying discussion.

File: core/src/main/scala/astraea/spark/rasterframes/RasterFunctions.scala (hunk 1 above)
Comment (vpipkt) [reply core/src/main/scala/astraea/spark/rasterframes/RasterFunctions.scala]: I've removed

**Comment by vpipkt** [comment]
Ok I think we are set up for making a final decision on naming. Right now it is `rf_for_all` and `rf_exists`. Pretty easy to change.

**Review by metasim — APPROVED** [review]

**Pull request merged**